clash-cores/src/Clash/Cores/Xilinx/DcFifo/BlackBoxes.hs

```diff
@@ -0,0 +1,215 @@
+{-|
+  Copyright   :  (C) 2022 Google Inc
+  License     :  BSD2 (see the file LICENSE)
+  Maintainer  :  QBayLogic B.V. <devops@qbaylogic.com>
+
+  Blackbox implementation for primitives in "Clash.Cores.Xilinx.DcFifo.Explicit".
+-}
+
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ViewPatterns #-}
+
+module Clash.Cores.Xilinx.DcFifo.BlackBoxes where
+
+import Prelude
+
+import Clash.Core.Literal (Literal(NaturalLiteral))
+import Clash.Core.TermLiteral (termToDataError, TermLiteralSNat(..))
+import Clash.Core.Term (Term(Literal))
+import qualified Clash.Primitives.DSL as DSL
+import Clash.Netlist.BlackBox.Types (BlackBoxFunction, emptyBlackBoxMeta)
+import Clash.Netlist.Types (TemplateFunction(..), BlackBox(BBFunction))
+import Clash.Netlist.Util (orNothing)
+
+import Clash.Netlist.Id (Identifier)
+import qualified Clash.Netlist.Id as Id
+import qualified Clash.Netlist.Types as N
+import qualified Clash.Netlist.BlackBox.Types as N
+
+import Control.Monad (unless)
+import Data.Either (lefts)
+import Data.Maybe (catMaybes)
+import GHC.Stack (HasCallStack)
+
+import Clash.Cores.Xilinx.DcFifo.Explicit (DcConfig(..))
+import Clash.Cores.Xilinx.DcFifo.Instances ()
+import Clash.Cores.Xilinx.Common (toTclBool, renderTcl, defIpConfig, IpConfig (properties))
+
+-- | Blackbox function for 'Clash.Cores.Xilinx.Fifo.dcFifo'. It parses the "DcConfig"
+-- supplied to 'dcFifo' from its Term representation, and passes them to two
+-- template functions:
+--
+--   * 'dcFifoTclTF': renders TCL file calling Xilinx's `create_ip`
+--   * 'dcFifoTF': instantiates IP generated in 'dcFifoTclTF'
+--
+-- Additionally, it generates an unique module name for the Xilinx IP.
+--
+dcFifoBBF :: HasCallStack => BlackBoxFunction
+dcFifoBBF _isD _primName args _resTys = do
+  let
+    [  knownNatN, _knownDomainWrite, _knownDomainRead, _knownNatDepth
+     , _constraint1, _constraint2
+     , either error id . termToDataError -> dcConfig
+     , _wClk, _rClk, _rst, _wData
+     , _rEnable
+
+    -- TODO: Make this blackbox return multiple results, instead of a tuple. See:
+    --       https://github.com/clash-lang/clash-compiler/pull/1560
+    --  , _, _, _, _, _, _, _
+     ] = lefts args
+
+  dcFifoName <- Id.makeBasic "dcfifo"
+
+  let depth =
+        case knownNatN of
+          Literal (NaturalLiteral n) -> fromInteger n
+          _ -> error "Unexpected type of knownNatN!"
```

Comment:
We don't end error messages in exclamation marks. And it sounds like the check should be done in the `TemplateFunction`, returning `TemplateFunction used valid tf` with `valid` returning `False` instead.

Also, note that you can get the `width` from the `BlackBoxContext` in your `tf` from `TemplateFunction used valid tf`. But then you'd have to do it twice, so I think it's fine to do it here, just don't throw the error here please, use the validation function.

Reply:
Doing the valid stuff kinda sucks, because you can't attach any error message :(.

Reply:
It's an internal error anyway, though, right? It means a mismatch between the Haskell primitive and its black box. I don't think a user would be any the wiser with the error message _Unexpected type of knownNatN_. If somebody can figure out the mistake based on that error message, they're probably clever enough to figure out the validation function as well.

Reply:
If this is not precisely what the validation function is for, what is it for?

Reply:
No, I just think we should rework it such that it returns a `Maybe ErrorMessage`, instead of a `Bool`. A lot of the time you want to say something along the lines of: "Did you pass in a constant? If you did, consider using TH, bla bla".

Reply:
Ideally you'd use it as some kind of "parse" function, so the validation would return `Either ErrorMessage a`, where the `a` would be passed to the second stage of the blackbox.

Reply:
How about

```haskell
TemplateFunction :: 
  [Int] ->
  (forall s . Backend s => BlackBoxContext -> Either ErrorMessage (Lazy.State s Doc)) ->
  TemplateFunction
```

The advantage is you don't need to make some type `a` for `Either ErrorMessage a`, you can just pass them as arguments to a function that produces the `Backend s => Lazy.State s Doc`.

Reply:
Are you sure you need to do that? The error is in a `BlackBoxFunction` so I think that has `Either String ...` already.

I pushed 1138dbe921621b28b3988b610bbfdb53b73edf65 which has the approach I believe makes sense.

Reply:
Oh woops... Yep, that commit looks good.

Reply:
I can confirm this bubbles up as an error! So for now I will resolve this discussion.